cart.go

```diff
@@ -13,60 +13,131 @@ import (
+// It also handles some $foo without parens, but we avoid using that.
+func (e *Expander) Expand(src string) string {
+	return os.Expand(src, e.Get)
+}
+
 var (
-	project               string
-	branch                string
-	buildNum              int
-	circleToken           string
-	outputPath            string
-	workflowArtifactBuild string
-	workflowDepth         int
-	verbose, dryRun       bool
+	circleToken     string
+	filter          FilterSet
+	verbose, dryRun bool
 )
 
 func main() {
+	var (
+		project             string
+		buildNum            int
+		outputPath          string
+		retrieveBuildsCount int
+	)
+
 	log.SetFlags(log.Lshortfile)
 	log.SetOutput(os.Stderr)
 
-	flag.StringVar(&project, "repo", "", "github `username/repo`")
-	flag.StringVar(&branch, "branch", "master", "search builds for branch `name`")
-	flag.IntVar(&buildNum, "build", 0, "get artifact for build number, ignoring branch")
-	flag.StringVar(&workflowArtifactBuild, "workflow-artifact-build", "", "if using a workflow, look for artifacts under build: `name`")
-	flag.IntVar(&workflowDepth, "workflow-depth", defaultWorkflowDepth, "how many workflow builds to check")
 	flag.StringVar(&circleToken, "token", "", "CircleCI auth token")
 	flag.StringVar(&outputPath, "o", "", "output file `path`")
 	flag.BoolVar(&verbose, "v", false, "verbose output")
 	flag.BoolVar(&dryRun, "n", false, "skip artifact download")
 
+	flag.StringVar(&project, "repo", "", "github `username/repo`")
+	flag.IntVar(&buildNum, "build", 0, "get artifact for build number, ignoring branch")
+	flag.StringVar(&filter.branch, "branch", "master", "search builds for branch `name`")
+
+	// Workflows:
+	// If there are multiple workflows, then the latest "build" is perhaps unrelated to building,
+	// not even a later step in a workflow where an earlier step did build.  Eg, we have
+	// stuff to automate dependencies checking, scheduled from cron.
+	// So to retrieve an artifact, we want to only consider specific workflow names.
+	// HOWEVER: those are config items in `.circleci/config.yml` and we should avoid hardcoding
+	// such arbitrary choices across more than one repo, so our default for now is empty,
+	// thus not filtered.
+	//
+	// Within a workflow, the build might not be the last step in the flow; it usually won't be.
+	// Later steps might be "deploy", "stash image somewhere", etc.
+	// So we need to step back from the last step within a workflow until we find the specific
+	// step we're told.
+	//
+	// Eg, for one project, at this time, we use "commit_workflow" as the workflow to search for
+	// and "build" as the job within that workflow.
+
+	flag.StringVar(&filter.workflow, "workflow", "", "only consider builds which are part of this workflow")
+	flag.StringVar(&filter.jobname, "jobname", "", "look within workflow for artifacts from this job's name")
+	flag.IntVar(&retrieveBuildsCount, "retrieve-count", defaultRetrieveCount, "how many builds to retrieve")
```

Comment:
- [ ] `-retrieve-count` → `-count` and `-c`

Reply:
Both that and my name are wrong, because they both imply retrieving N artifacts.  We only ever retrieve from one build.  I'm going with `-search-depth` and no short option because it's not a routine-usage flag.  If we need to override this, then we should change the code default shortly afterwards.